src/lib/Syntax.hs

```diff
@@ -77,10 +82,11 @@ data Atom = Var Var
           | Pi  PiType
           | DataCon DataDef [Atom] Int [Atom]
           | TypeCon DataDef [Atom]
+          | LabeledRow (ExtLabeledItems Type Name)
           | Record (LabeledItems Atom)
```

Comment:
Hmm why do regular records don't use `ExtLabeledItems`? Is it because you can't have a run-time representation of a record unless you know all of its fields, unlike for a variant?

Reply:
Yeah, that's exactly right.

Reply:
Great. I wonder if we could do the same for variants. I understand that inference is no longer quite as straightforward, since you don't know the type exactly when you see the constructor, but I suspect that it should be fully known after type inference right? This way we could make `Atom`s be fully specified as all ambiguity would be resolved at the point where we translate the surface language (`UExpr`) into core IR, right?

Reply:
That's an interesting thought. We would still need something to represent a variant with a not-fully-known type in the IR, though, so I guess that would have to be an expr or something? So we would have
```
data Atom = ...
  | Variant (LabeledItems Type) Label Int Atom

...

data Expr = ...
  | AmbiguousVariant (ExtLabeledItems Type Name) Label Int Atom
```
or similar?

I'm not sure this really adds anything, though? (Related note: it might still be unknown until function inlining, in the case where the tail is an implicit function argument, so it could last at least until simplification for evaluated things and forever inside function bodies.)

Reply:
Also, `RecordTy` and `VariantTy` are still ambiguous. The first argument to the `Variant` constructor is basically just it's type, since you can't infer the type of a variant from its contents. (I guess a different design decision would be to make it just `Variant Type Label Int Atom` where the first argument is always a `VariantTy` atom? But that seemed unnecessarily indirect.)

Reply:
Another related point here is that `Variant` atoms with a tail `Ext _ (Just tail)` transform into unambiguous fully-specified `Variant` atoms during substitution, which is convenient. Records can't do that because a record atom has to actually have values for all of its fields. In a sense the only things that it makes sense to put in an `ExtLabeledItems` are types, not values. Maybe that's the more fundamental reason why `Record` just has a `LabeledItems` (and the equivalent of an extended record value is just `RecordCons`).